File: Udacity/Term3/NaiveBayes/Cpp/classifier.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <cmath>
#include <vector>
#include "classifier.h"

double gaussians_calc(double x, double mu, double sigma)
{
    return exp(-((x - mu)*(x - mu))/(2*sigma*sigma)) / (sigma * sqrt(2*M_PI));
}

int find_pos(vector<string> array, string x)
{
    for(int i=0; i<array.size(); i++)
    {
        if(array[i] == x)
        {
            return i;
        }
    }
    return -1;
}


/**
 * Initializes GNB
 */
GNB::GNB() {
    N_features = 4;
    N_labels = 3;
}

GNB::~GNB() {}

void GNB::train(vector<vector<double> > data, vector<string> labels)
{
    // this-> possible_labels =  array_uniq(labels)
    this-> N_features = data[0].size();
    this-> N_labels = this-> possible_labels.size();
    N_labels = this-> N_labels;
    N_features = this-> N_features;
    double label_sum [N_labels][N_features];
    vector<int> count_label(3, 0);
    // init.
    for(int i=0; i < N_labels; i++)
    {
        for(int j=0; j < N_features; j++)
        {
            label_sum[i][j] = 0;
            this->std[i][j] = 0;
        }
    }

    // calc sum first.
    for(int i=0; i < data.size(); i++)
    {
        int pos = find_pos(this-> possible_labels, labels[i]);
        count_label[pos] += 1;
        for(int j=0; j<N_features; j++)
        {
            
            label_sum[pos][j] += data[i][j];
        }
    }   
    // calc mean
    for(int i=0; i<N_labels; i++)
    {
        for(int j=0; j<N_features;  j++)
        {
            this->mean[i][j] = (double)label_sum[i][j] / data.size();
            //cout << "mean: " << mean[i][j] << endl;
        }
    }
    // calc std
    for(int i=0; i < data.size(); i++)
    {
        int pos = find_pos(this-> possible_labels, labels[i]);
        double delta;
        for(int j=0; j<N_features; j++)
        {
            delta = (data[i][j] - this->mean[pos][j]);
            this->std[pos][j] +=  delta * delta;
        }

    }   
    for(int i=0; i<N_labels; i++)
    {
        for(int j=0; j<N_features; j++)
        {
            double tmp = this->std[i][j];
            this->std[i][j] = sqrt(tmp / count_label[i]);
        }
    }

}

string GNB::predict(vector<double> sample)
{
    /*
        Once trained, this method is called and expected to return 
        a predicted behavior for the given observation.

        INPUTS

        observation - a 4 tuple with s, d, s_dot, d_dot.
          - Example: [3.5, 0.1, 8.5, -0.2]

        OUTPUT

        A label representing the best guess of the classifier. Can
        be one of "left", "keep" or "right".
        """
        # TODO - complete this
    */
    double max_possibility = 0;
    double possibility;
    int max_possibility_pos;
    for(int i=0; i< this->N_labels; i++)
    {
        possibility = 0;
        for(int j=0; j < this-> N_features; j++)
        {
            possibility += gaussians_calc(sample[j], mean[i][j], std[i][j]);
            
        }
        //cout << possibility << endl;
        if(possibility > max_possibility)
        {
            max_possibility = possibility;
            max_possibility_pos = i;
        }
    }

    //cout << this->possible_labels[max_possibility_pos] << endl;
    return this->possible_labels[max_possibility_pos];

}

File: Udacity-Term3/NaiveBayes/python/classifier.py
# *_ coding: UTF-8 _*
# 1. 给每个label计算每个feature的高斯函数（每个feature有 N个高斯函数）
# 2. 给定一个train，计算各feature的N个高斯函数对应的概率
# 	 把各feature代表的相同label相乘，得到最大概率的label 
from pandas.core.frame import DataFrame
import numpy as np

class GNB(object):

	def __init__(self):
		self.possible_labels = ['left', 'keep', 'right']

		self.gaussians = {}
	
	def gaussians_calc(self, x,mu,sigma):
		return np.exp(-((x - mu)**2)/(2*sigma**2)) / (sigma * np.sqrt(2*np.pi))

	def train(self, data, labels):
		"""
		Trains the classifier with N data points and labels.

		INPUTS
		data - array of N observations
		  - Each observation is a tuple with 4 values: s, d, 
		    s_dot and d_dot.
		  - Example : [
			  	[3.5, 0.1, 5.9, -0.02],
			  	[8.0, -0.3, 3.0, 2.2],
			  	...
		  	]

		labels - array of N labels
		  - Each label is one of "left", "keep", or "right".
		"""
		data = DataFrame(data)
		data['labels'] = labels
		#print data.head()
		for label in self.possible_labels:
			df = data.loc[(data['labels'] == label)]
			mean_list = np.array(df.mean())
			sum_list = np.array(df.sum())
			std_list = np.array(df.std())
			self.gaussians[label] = [zip(mean_list, std_list)]
			print mean_list
	def predict(self, observation):
		"""
		Once trained, this method is called and expected to return 
		a predicted behavior for the given observation.

		INPUTS

		observation - a 4 tuple with s, d, s_dot, d_dot.
		  - Example: [3.5, 0.1, 8.5, -0.2]

		OUTPUT

		A label representing the best guess of the classifier. Can
		be one of "left", "keep" or "right".
		"""
		# TODO - complete this
		p_max = 0
		predict = None
		for label in self.possible_labels:
			possiblity = 1
			label_gaussian = self.gaussians.get(label)
			for i in range(len(observation)):
				(mean, std) = label_gaussian[0][i]
				value = observation[i]
				possiblity *= self.gaussians_calc(value, mean, std)
			if p_max < possiblity:
				p_max = possiblity
				predict = label

		return predict

